pkg/catalog/utils/version.go

```diff
@@ -63,6 +65,41 @@ func VersionGreaterThan(a, b string) bool {
 	return version.GreaterThan(a, b)
 }
 
+func ValidateChartCompatibility(template *v3.CatalogTemplateVersion, clusterlister v3.ClusterLister, clusterName string) error {
+	if err := ValidateRancherVersion(template); err != nil {
+		return err
+	}
+	if err := ValidateKubeVersion(template, clusterlister, clusterName); err != nil {
+		return err
+	}
+	return nil
+}
+
+func ValidateKubeVersion(template *v3.CatalogTemplateVersion, clusterlister v3.ClusterLister, clusterName string) error {
+	if template.Spec.KubeVersion == "" {
+		return nil
+	}
+	constraint, err := semver.ParseRange(template.Spec.KubeVersion)
+	if err != nil {
+		logrus.Errorf("failed to parse constraint for kubeversion %s: %v", template.Spec.KubeVersion, err)
+		return nil
+	}
+
+	cluster, err := clusterlister.Get("", clusterName)
+	if err != nil {
+		return err
+	}
+
+	k8sVersion, err := semver.Parse(strings.TrimPrefix(cluster.Status.Version.String(), "v"))
+	if err != nil {
+		return err
+	}
+	if !constraint(k8sVersion) {
+		return httperror.NewAPIError(httperror.InvalidBodyContent, fmt.Sprintf("incompatible kubernetes version [%s] for template [%s]", k8sVersion.String(), template.Name))
+	}
+	return nil
+}
+
```

Comment:
Trying to wrap my head around the flows going on here, can you summarize which methods are primary ones and how they are being used. Somewhat confusing as there are several with the same name.

Reply:
This method, ValidateKubeVersion, in the utils package is the main one being added. Before, most charts were just being checked for their compatibility with the Rancher server version with a method called ValidateRancherVersion. 

Now both of these methods, ValidateKubeVersion and ValidateRancherVersion are in a new method called ValidateChartCompatibility. Most occurrences of ValidateRancherVersion have been replaced with this function. Any other functions called validateChartCompatibility are helper functions in their own respective packages that prepare their data for the functions and do any additional processing. This is because compatibility is validated in stores, validators, controllers, and action handlers which may all vary a bit in how they extract the information needed.